arex-schedule-web-api/src/main/java/com/arextest/schedule/service/ReplayCaseTransmitService.java

```diff
@@ -55,61 +58,73 @@ public class ReplayCaseTransmitService {
+        Map<ReplayActionItem, List<ReplayActionCaseItem>> actionsOfBatch = caseItems.stream()
+                .collect(Collectors.groupingBy(ReplayActionCaseItem::getParent));
+
+        actionsOfBatch.forEach((actionItem, casesOfAction) -> {
+                    // warmUp should be done once for each endpoint
+                    if (!actionItem.isItemProcessed()) {
+                        actionItem.setItemProcessed(true);
+                        progressEvent.onActionBeforeSend(actionItem);
+                        // todo add biz log
+                        activeRemoteHost(casesOfAction);
+                    }
+                });
+    }
+
+    public void releaseCasesOfContext(ReplayPlan replayPlan, PlanExecutionContext<?> executionContext) {
         // checkpoint: before skipping batch of group
         if (executionContext.getExecutionStatus().isAbnormal()) {
             return;
         }
 
-        int contextCasesCount = (int) replayActionCaseItemRepository.countWaitingSendList(replayActionItem.getId(),
+        Map<String, Long> caseCountMap = replayActionCaseItemRepository.countWaitHandlingByAction(replayPlan.getId(),
                 executionContext.getContextCaseQuery());
+        Map<String, ReplayActionItem> actionItemMap = replayPlan.getActionItemMap();
 
-        replayActionItem.recordProcessCaseCount(contextCasesCount);
-        replayActionItem.getSendRateLimiter().batchRelease(false, contextCasesCount);
+        for (Map.Entry<String, Long> actionIdToCaseCount : caseCountMap.entrySet()) {
+            ReplayActionItem replayActionItem = actionItemMap.get(actionIdToCaseCount.getKey());
+            int contextCasesCount = actionIdToCaseCount.getValue().intValue();
+            replayActionItem.recordProcessCaseCount(contextCasesCount);
+            replayActionItem.getSendRateLimiter().batchRelease(false, contextCasesCount);
 
-        // if we skip the rest of cases remaining in the action item, set its status
-        if (replayActionItem.getReplayCaseCount() == replayActionItem.getCaseProcessCount().get()) {
-            progressEvent.onActionInterrupted(replayActionItem);
-            ReplayPlan replayPlan = replayActionItem.getParent();
-            for (int i = 0; i < contextCasesCount; i++) {
-                progressTracer.finishCaseByPlan(replayPlan);
-            }
-        } else {
-            for (int i = 0; i < contextCasesCount; i++) {
-                progressTracer.finishCaseByAction(replayActionItem);
+            // if we skip the rest of cases remaining in the action item, set its status
+            if (replayActionItem.getReplayCaseCount() == replayActionItem.getCaseProcessCount().intValue()) {
+                progressEvent.onActionInterrupted(replayActionItem);
+                for (int i = 0; i < contextCasesCount; i++) {
+                    progressTracer.finishCaseByPlan(replayPlan);
```

Comment:
Why does It need to repeat the same operation?

Reply:
each call to the progressTracer.finishCaseByPlan() method will reduce the redis plan case counter by 1,
if we skip 100 case, we will need to call it 100 times,
Batch redis operation is not currently supported by Arex-Common package